packages/leancode_lint/lib/custom_lints/add_sliver_prefix_for_widget_returning_sliver.dart

```diff
@@ -0,0 +1,145 @@
+// ignore_for_file: comment_references
+
+import 'package:analyzer/dart/ast/ast.dart';
+import 'package:analyzer/error/listener.dart';
+import 'package:collection/collection.dart';
+import 'package:custom_lint_builder/custom_lint_builder.dart';
+
+/// Displays warning for widgets returning slivers but not having [Sliver] or
+/// [_Sliver] prefix.
+class AddSliverPrefixForWidgetReturningSliver extends DartLintRule {
+  AddSliverPrefixForWidgetReturningSliver() : super(code: _getLintCode());
+
+  static const ruleName = 'add_sliver_prefix_for_widget_returning_sliver';
+
+  @override
+  void run(
+    CustomLintResolver resolver,
+    ErrorReporter reporter,
+    CustomLintContext context,
+  ) {
+    context.registry.addClassDeclaration(
+      (node) {
+        final isWidgetClass = _isWidgetClass(node);
+        if (!isWidgetClass) {
+          return;
+        }
+
+        final startsWithSliver = _hasSliverPrefix(node.name.lexeme);
+        if (startsWithSliver) {
+          return;
+        }
+
+        final buildMethod = _getBuildMethod(node);
+        if (buildMethod == null) {
+          return;
+        }
+
+        // Get all return expressions from build method
+        final returnExpressions = switch (buildMethod.body) {
+          ExpressionFunctionBody(:final expression) => [expression],
+          BlockFunctionBody(:final block) =>
+            block.statements.expand(_getAllInnerReturnStatements).toList(),
+          _ => <Expression>[],
+        };
+
+        final isSliver = _anyReturnsSliver(returnExpressions);
+
+        if (node case ClassDeclaration(:final declaredElement?) when isSliver) {
+          reporter.reportErrorForElement(
+            _getLintCode(node.name.lexeme),
+            declaredElement,
+          );
+        }
+      },
+    );
+  }
+
+  /// Returns all return expressions from passed statement recursively.
+  Iterable<Expression> _getAllInnerReturnStatements(Statement statement) {
+    switch (statement) {
+      case IfStatement():
+        return [
+          ..._getAllInnerReturnStatements(statement.thenStatement),
+          if (statement.elseStatement case final statement?)
+            ..._getAllInnerReturnStatements(statement),
+        ];
+
+      case ForStatement():
+        return _getAllInnerReturnStatements(statement.body);
```

Comment:
```suggestion
      case ForStatement(:final body):
        return _getAllInnerReturnStatements(body);
```
Similarly for other cases